### PR TITLE
Added attribute for custom component names in modules

### PR DESCRIPTION
**Pull request by shaltielshmid**
Added an attribute to allow the users to specify a custom component name instead of the default behavior of using the field name. 
This can be very helpful when using models from PyTorch, and wanting to maintain the naming conventions in each language whilst sharing the state_dict parameter names. 

For example, in python I would write:

```python
self.word_embeddings = nn.Embeddings(30522, 768);
```

And in C# I would write:
```cs
[ComponentName(Name = "word_embeddings")]
private readonly Embedding _wordEmbeddings;
```

This way I can use the C# naming convention but also not have to worry about renaming the parameters when I save the python state_dict.

### CODE REVIEW

**Comment by shaltielshmid** [comment]
@dotnet-policy-service agree

**Comment by shaltielshmid** [comment]
I see that the CI failed, on two items:
`Windows_x64_NetFX Release_Build`: All the tests it ran passed, but it seems to have timed out on a network connection, the error is:
```Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host..```
How can I rerun the CI? 

`Ubuntu_x64 Debug_Build`: Azure seemed to indicate that this test has been failing a lot recently, is this possibly a different issue unrelated to my changes?

File: src/TorchSharp/Utils/ComponentNameAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TorchSharp.Utils
+{
+    public class ComponentNameAttribute : Attribute
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Utils/ComponentNameAttribute.cs:9]: Doc comments, please!

Comment (shaltielshmid) [reply src/TorchSharp/Utils/ComponentNameAttribute.cs:12]: Added!

File: src/TorchSharp/NN/Module.cs [hunk 2]
```diff
@@ -1083,7 +1084,7 @@ protected virtual void RegisterComponents()
 
                     foreach (var field in GetType().GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)) {
 
-                        var fieldName = field.Name;
+                        var fieldName = field.GetCustomAttribute<ComponentNameAttribute>()?.Name ?? field.Name;
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/Module.cs:1087]: There are 3 other places where the code goes through the fields and does stuff, specifically, when moving a module to/from GPU, or converting to another element type. They need the same treatment.

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]
Please make sure that all the places where the explicit name is needed are covered.

**Review by NiklasGustafsson — DISMISSED** [review]

**Comment by NiklasGustafsson** [comment]
I will merge this after the 2.1.0 work that I've been doing, but it will go in before the next release later this week.

**Comment by shaltielshmid** [comment]
Great! Thanks!

**Comment by NiklasGustafsson** [comment]
@shaltielshmid -- unit tests, too...

**Comment by NiklasGustafsson** [comment]
Add a line to the release notes under the v0.101.0 heading / API Changes.

**Comment by shaltielshmid** [comment]
Sure, done. 
Added unit tests & release notes.

**Review by NiklasGustafsson — APPROVED** [review]

**Pull request merged**